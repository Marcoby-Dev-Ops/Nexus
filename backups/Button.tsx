import React from 'react';
import { cn, getButtonClasses, buttonVariants, buttonSizes } from '@/lib/styles';
import { Spinner } from './Spinner';
import PropTypes from 'prop-types';

/**
 * @interface ButtonProps
 * @description Props for the Button component.
 */
export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: keyof typeof buttonVariants | 'default';
  size?: keyof typeof buttonSizes | 'default';
  isLoading?: boolean;
}

/**
 * @name Button
 * @description A reusable button component with variants and loading state.
 * @param {ButtonProps} props - The props for the component.
 * @returns {JSX.Element} The rendered Button component.
 */
export const Button: React.FC<ButtonProps> = ({
  children,
  className,
  variant = 'primary',
  size = 'md',
  isLoading,
  disabled,
  ...props
}) => {
  // Map 'default' to 'primary' for compatibility
  const mappedVariant = variant === 'default' ? 'primary' : variant as keyof typeof buttonVariants;
  const mappedSize = size === 'default' ? 'md' : size as keyof typeof buttonSizes;
  
  return (
    <button
      className={getButtonClasses(mappedVariant, mappedSize, className)}
      disabled={disabled || isLoading}
      {...props}
    >
      {isLoading ? (
        <div className="mr-2">
          <Spinner size={16} className="text-current" />
        </div>
      ) : null}
      {children}
    </button>
  );
};

Button.displayName = 'Button';

Button.propTypes = {
  variant: PropTypes.oneOf(['primary', 'secondary', 'destructive', 'ghost', 'link', 'outline'] as const),
  size: PropTypes.oneOf(['sm', 'md', 'lg', 'icon', 'iconSm', 'iconLg'] as const),
  isLoading: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
};

export default Button; 