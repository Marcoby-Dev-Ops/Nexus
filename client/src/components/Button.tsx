import React from 'react';
import PropTypes from 'prop-types';

/**
 * @interface ButtonProps
 * @description Props for the Button component.
 */
export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'primary' | 'secondary' | 'destructive';
  loading?: boolean;
}

/**
 * @name Button
 * @description A reusable button component with variants and loading state.
 * @param {ButtonProps} props - The props for the component.
 * @returns {JSX.Element} The rendered Button component.
 */
export const Button: React.FC<ButtonProps> = React.memo(({ 
  variant = 'primary', 
  loading = false, 
  disabled, 
  children, 
  className = '', 
  ...props 
}) => {
  const base = 'inline-flex items-center justify-center rounded-md px-4 py-2 font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none';
  const variants = {
    primary: 'bg-primary text-primary-foreground hover:bg-primary/90',
    secondary: 'bg-secondary text-secondary-foreground hover:bg-secondary/80',
    destructive: 'bg-destructive text-destructive-foreground hover:bg-destructive/80',
  };
  return (
    <button
      className={`${base} ${variants[variant]} ${className}`}
      disabled={disabled || loading}
      {...props}
    >
      {loading && (
        <svg className="animate-spin mr-2 h-4 w-4 text-current" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
          <circle className="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" strokeWidth="4" />
          <path className="opacity-75" fill="currentColor" d="M4 12a8 8 0 018-8v8z" />
        </svg>
      )}
      {children}
    </button>
  );
});

Button.displayName = 'Button';

Button.propTypes = {
  variant: PropTypes.oneOf(['primary', 'secondary', 'destructive']),
  loading: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
};

export default Button; 