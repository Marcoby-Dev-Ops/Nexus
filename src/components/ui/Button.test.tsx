/**
 * @file Button.test.tsx
 * @description Unit tests for the Button component using React Testing Library and Jest.
 */
import '@testing-library/jest-dom';
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Button } from './Button';

describe('Button', () => {
  it('renders with children', () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText(/click me/i)).toBeInTheDocument();
  });

  it('calls onClick when clicked', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click me</Button>);
    fireEvent.click(screen.getByText(/click me/i));
    expect(handleClick).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Button>Snapshot Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
}); 