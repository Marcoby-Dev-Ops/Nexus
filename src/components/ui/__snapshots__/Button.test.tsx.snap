// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 focus-visible:ring-ring shadow-sm h-10 px-4"
  >
    Snapshot Button
  </button>
</DocumentFragment>
`;
